Added debug option to log loading and renaming

dcamilleri opened the pull request:
Hi Jack,

Saw your tweet and issue #90  and added some logs for loading and renaming packages. Also added chalk for fancy colors (used green for loading and yellow for renaming). Didn't add tests given that it's just console logs.
Let me know if you'd like anything else to be logged :) !

Cheers,
Dorian.

callumacrae commented:
Looks good to me :+1:

Perhaps split the lazy loading logging into two? One for when the property is added to the object, and one when it is required.

jackfranklin commented on index.js:
```diff
@@ -20,6 +21,7 @@ module.exports = function(options) {
   var requireFn;
   options = options || {};
 
+  var DEBUG = options.DEBUG;
```
can we have an explicit `|| false` here just to be clear? (Pedantic, I know).

jackfranklin commented on index.js:
```diff
@@ -58,14 +60,24 @@ module.exports = function(options) {
+  function logDebug(message, chalkColor) {
+    console.log(chalkColor(message));
+  }
+
```
I would move the `if(DEBUG)` into this function:

```
function logDebug(...) {
  if (DEBUG) { console.log(...) }
}
```

Then we don't have to have the `if (DEBUG)` repeated every time we want to log something

jackfranklin commented:
> Perhaps split the lazy loading logging into two? One for when the property is added to the object, and one when it is required.

I'd agree with that.

I'd also like one that just logs out the entire array of plugins that were found, and one maybe that logs out all the configured options in an object (as JSON perhaps)?

Left a couple of comments, but other than those it's great. Thank you so much for doing this, you're awesome. Make sure to also edit the package.json file and add yourself as a contributor.

shinnn commented:
There is an official logger library [used in gulp 4](https://github.com/gulpjs/gulp-cli/pull/32), [`gulplog`](https://github.com/gulpjs/gulplog), by which users can set the application-wide log level (e.g. `debug`, `info`, `warn`).

Though I appreciate @dcamilleri's great work, I'd like to suggest:

1. Wait for the stable release of gulp 4.0.0.
2. Implement this feature by using `gulplog`, instead of using bare `console.log`.

callumacrae commented:
You don't need to wait until gulp 4 to use gulplog, iirc

shinnn commented:
> You don't need to wait until gulp 4 to use gulplog, iirc

Oh, thanks for letting me know.

jackfranklin commented:
The only thing with that though is it would be impossible to turn logging on or off specifically for gulp-load-plugins? I'd quite like that ability - I'm happy to merge with `console.log`. @callumacrae what do you think?

callumacrae commented:
You would still use `if(DEBUG) {`, just wrapped around the gulplog calls, not `console.log`.

dcamilleri commented:
Thank you all for your feedback, I really appreciate it.

Pushed another commit with the list of plugins found and configured options. Also split the lazy loading into two and updated the package.json.

Let me know if that's correct!

jackfranklin commented:
Ahh sorry I misunderstood! Let's swap to using `gulplog` and use `logger.info` for now. Then we should be done!

callumacrae commented:
Also, how about moving the `gulp-load-plugins:` prefix into `logDebug`?

Could also make the colour argument default to `chalk.green`.

callumacrae commented:
What's the advantage of having colour as an argument vs just calling it? This way, you can't _not_ specify a colour. I reckon either the function should accept a string like `green` which it calls on the chalk object, or just not deal with the colour argument.

dcamilleri commented:
@callumacrae Are you sure `gulplog` can be used before 4.0.0 ? It doesn't seem to work on my end (using gulp@3.9.0). Maybe we can use [has-gulplog](https://github.com/gulpjs/has-gulplog) to detect it, and if not present use `console.log`.

Also totally agree about moving the prefix to `logDebug` function. Will fix it in the next commit.

callumacrae commented:
Hmm, that's odd. @phated Does gulplog only work on gulp 4.0? Won't this break backward compatibility with plugins?

phated commented:
Gulplog can be used with the 4.0 branch of gulp-cli

I recommend using gulp-util until we release 1.0 of gulp-cli

callumacrae commented:
Awesome, thanks :smiley:

dcamilleri commented:
Great thanks !
I implemented it using `gulp-util`, and added green as the default color, as asked.

phated commented:
Btw, if this module is used outside of gulp, then gulplog won't work in those situations.  gulp-cli will be aware of the gulplog event emitter and know how to print the messages.

jackfranklin commented:
@phated I'm kind of fine with that; it won't break but it just won't log, which isn't that big of a deal. I don't really want to encourage people to use this plugin more generally for now anyway.

@dcamilleri awesome work :+1: I'll take a proper look in the morning with fresh eyes and then release as 1.0.1.

callumacrae commented:
:tada:

dcamilleri commented:
:tada: Thanks a lot for your patience and your feedback guys !

jackfranklin commented:
@dcamilleri thanks for your work implementing! It's out in 1.1.0.
